File: api/websocket.js
// WebSocket Server using Socket.IO
// Save this as: /api/websocket.js

const jwt = require('jsonwebtoken');
const axios = require('axios');

function setupWebSocket(server) {
    // Initialize Socket.IO
    const io = require('socket.io')(server, {
        cors: {
            origin: process.env.NODE_ENV === 'production' 
                ? ['https://mstymillionaire.com', 'https://www.mstymillionaire.com']
                : ['http://localhost:3000', 'http://127.0.0.1:3000'],
            credentials: true
        },
        transports: ['websocket', 'polling']
    });

    // Track subscriptions
    const userSubscriptions = new Map();
    const tickerSubscribers = new Map();

    // Authentication middleware
    io.use(async (socket, next) => {
        try {
            const token = socket.handshake.auth.token;
            
            if (!token) {
                return next(new Error('Authentication required'));
            }

            // Verify JWT token
            const decoded = jwt.verify(token, process.env.JWT_SECRET);
            socket.userId = decoded.userId;
            socket.userEmail = decoded.email;
            
            next();
        } catch (err) {
            next(new Error('Invalid token'));
        }
    });

    // Handle connections
    io.on('connection', (socket) => {
        console.log(`User ${socket.userId} connected - Socket ID: ${socket.id}`);
        
        // Initialize user's subscription set
        if (!userSubscriptions.has(socket.userId)) {
            userSubscriptions.set(socket.userId, new Set());
        }

        // Send authentication success
        socket.emit('auth_success', {
            userId: socket.userId,
            socketId: socket.id
        });

        // Handle ticker subscription
        socket.on('subscribe', (data) => {
            const { ticker } = data;
            if (!ticker) return;

            const tickerUpper = ticker.toUpperCase();
            
            // Add to user's subscriptions
            userSubscriptions.get(socket.userId).add(tickerUpper);
            
            // Add to ticker's subscribers
            if (!tickerSubscribers.has(tickerUpper)) {
                tickerSubscribers.set(tickerUpper, new Set());
            }
            tickerSubscribers.get(tickerUpper).add(socket.id);
            
            // Join room for this ticker
            socket.join(`ticker:${tickerUpper}`);
            
            console.log(`User ${socket.userId} subscribed to ${tickerUpper}`);
            socket.emit('subscribed', { ticker: tickerUpper });
            
            // Send current price immediately
            sendCurrentPrice(socket, tickerUpper);
        });

        // Handle ticker unsubscription
        socket.on('unsubscribe', (data) => {
            const { ticker } = data;
            if (!ticker) return;

            const tickerUpper = ticker.toUpperCase();
            
            // Remove from user's subscriptions
            userSubscriptions.get(socket.userId).delete(tickerUpper);
            
            // Remove from ticker's subscribers
            if (tickerSubscribers.has(tickerUpper)) {
                tickerSubscribers.get(tickerUpper).delete(socket.id);
            }
            
            // Leave room
            socket.leave(`ticker:${tickerUpper}`);
            
            console.log(`User ${socket.userId} unsubscribed from ${tickerUpper}`);
            socket.emit('unsubscribed', { ticker: tickerUpper });
        });

        // Handle batch subscription
        socket.on('subscribe_batch', (data) => {
            const { tickers } = data;
            if (!Array.isArray(tickers)) return;

            tickers.forEach(ticker => {
                const tickerUpper = ticker.toUpperCase();
                userSubscriptions.get(socket.userId).add(tickerUpper);
                
                if (!tickerSubscribers.has(tickerUpper)) {
                    tickerSubscribers.set(tickerUpper, new Set());
                }
                tickerSubscribers.get(tickerUpper).add(socket.id);
                
                socket.join(`ticker:${tickerUpper}`);
            });

            console.log(`User ${socket.userId} subscribed to batch:`, tickers);
        });

        // Handle batch unsubscription
        socket.on('unsubscribe_batch', (data) => {
            const { tickers } = data;
            if (!Array.isArray(tickers)) return;

            tickers.forEach(ticker => {
                const tickerUpper = ticker.toUpperCase();
                userSubscriptions.get(socket.userId).delete(tickerUpper);
                
                if (tickerSubscribers.has(tickerUpper)) {
                    tickerSubscribers.get(tickerUpper).delete(socket.id);
                }
                
                socket.leave(`ticker:${tickerUpper}`);
            });

            console.log(`User ${socket.userId} unsubscribed from batch:`, tickers);
        });

        // Handle disconnection
        socket.on('disconnect', (reason) => {
            console.log(`User ${socket.userId} disconnected - Reason: ${reason}`);
            
            // Clean up subscriptions
            const userSubs = userSubscriptions.get(socket.userId);
            if (userSubs) {
                userSubs.forEach(ticker => {
                    if (tickerSubscribers.has(ticker)) {
                        tickerSubscribers.get(ticker).delete(socket.id);
                    }
                });
            }
        });

        // Handle errors
        socket.on('error', (error) => {
            console.error(`Socket error for user ${socket.userId}:`, error);
        });
    });

    // Function to send current price
    async function sendCurrentPrice(socket, ticker) {
        try {
            // You would fetch from your API/cache here
            const response = await axios.get(`http://localhost:${process.env.PORT}/api/stock/${ticker}`, {
                headers: { 'Authorization': `Bearer ${process.env.INTERNAL_API_KEY}` }
            });
            
            socket.emit('ticker_update', response.data);
        } catch (error) {
            console.error(`Failed to fetch current price for ${ticker}:`, error);
        }
    }

    // Function to broadcast price updates
    function broadcastPriceUpdate(ticker, data) {
        io.to(`ticker:${ticker}`).emit('ticker_update', {
            ticker,
            price: data.price,
            change: data.change,
            changePercent: data.changePercent,
            volume: data.volume,
            high: data.high,
            low: data.low,
            timestamp: new Date()
        });
    }

    // Function to send alerts
    function sendPriceAlert(userId, alertData) {
        const sockets = Array.from(io.sockets.sockets.values())
            .filter(socket => socket.userId === userId);
        
        sockets.forEach(socket => {
            socket.emit('price_alert', alertData);
        });
    }

    // Start price update loop (every 5 seconds)
    setInterval(async () => {
        // Get all active tickers
        const activeTickers = Array.from(tickerSubscribers.keys());
        
        if (activeTickers.length === 0) return;

        // Fetch updates for all tickers
        for (const ticker of activeTickers) {
            try {
                // Fetch from your API
                const response = await axios.get(`http://localhost:${process.env.PORT}/api/stock/${ticker}`, {
                    headers: { 'Authorization': `Bearer ${process.env.INTERNAL_API_KEY}` }
                });
                
                // Broadcast to all subscribers
                broadcastPriceUpdate(ticker, response.data);
            } catch (error) {
                console.error(`Failed to update ${ticker}:`, error.message);
            }
        }
    }, 5000); // Update every 5 seconds

    // Return io instance for external use
    return io;
}

module.exports = setupWebSocket;

File: src/msty-millionaire-backend/src/routes/websocket.py
from flask import Blueprint
from flask_socketio import SocketIO, emit, join_room, leave_room
from flask_jwt_extended import jwt_required, get_jwt_identity, decode_token
import json
import threading
import time
from datetime import datetime
from src.routes.api import fetch_stock_data

# WebSocket blueprint
websocket_bp = Blueprint('websocket', __name__)

# Global SocketIO instance (will be initialized in main.py)
socketio = None

# Active subscriptions: {room_name: {user_id: socket_id}}
active_subscriptions = {}

# Background task for real-time data updates
background_thread = None
background_thread_started = False

def init_socketio(app):
    """Initialize SocketIO with the Flask app"""
    global socketio
    socketio = SocketIO(app, cors_allowed_origins="*", async_mode='eventlet')
    
    # Register event handlers
    @socketio.on('connect')
    def handle_connect(auth):
        """Handle client connection"""
        try:
            # Verify JWT token
            token = auth.get('token') if auth else None
            if not token:
                print("WebSocket connection rejected: No token provided")
                return False
            
            # Decode token to get user ID
            try:
                decoded_token = decode_token(token)
                user_id = decoded_token['sub']
                print(f"WebSocket connected: User {user_id}")
                
                # Store user info in session
                from flask import session
                session['user_id'] = user_id
                
                emit('connected', {'status': 'success', 'message': 'Connected to real-time data feed'})
                
                # Start background thread if not already running
                global background_thread_started
                if not background_thread_started:
                    start_background_thread()
                
                return True
                
            except Exception as e:
                print(f"WebSocket token validation failed: {e}")
                return False
                
        except Exception as e:
            print(f"WebSocket connection error: {e}")
            return False
    
    @socketio.on('disconnect')
    def handle_disconnect():
        """Handle client disconnection"""
        try:
            from flask import session
            user_id = session.get('user_id')
            if user_id:
                # Remove user from all subscriptions
                for room_name in list(active_subscriptions.keys()):
                    if user_id in active_subscriptions[room_name]:
                        del active_subscriptions[room_name][user_id]
                        if not active_subscriptions[room_name]:
                            del active_subscriptions[room_name]
                
                print(f"WebSocket disconnected: User {user_id}")
        except Exception as e:
            print(f"WebSocket disconnect error: {e}")
    
    @socketio.on('subscribe')
    def handle_subscribe(data):
        """Handle subscription to ticker updates"""
        try:
            from flask import session
            user_id = session.get('user_id')
            if not user_id:
                emit('error', {'message': 'Authentication required'})
                return
            
            ticker = data.get('ticker', '').upper()
            if not ticker:
                emit('error', {'message': 'Ticker is required'})
                return
            
            room_name = f"ticker_{ticker}"
            
            # Join the room
            join_room(room_name)
            
            # Track subscription
            if room_name not in active_subscriptions:
                active_subscriptions[room_name] = {}
            active_subscriptions[room_name][user_id] = True
            
            print(f"User {user_id} subscribed to {ticker}")
            emit('subscribed', {'ticker': ticker, 'status': 'success'})
            
            # Send current data immediately
            try:
                current_data = fetch_stock_data(ticker)
                emit('ticker_update', {
                    'ticker': ticker,
                    'data': current_data,
                    'timestamp': datetime.now().isoformat()
                }, room=room_name)
            except Exception as e:
                print(f"Error sending current data for {ticker}: {e}")
                
        except Exception as e:
            print(f"Subscribe error: {e}")
            emit('error', {'message': 'Subscription failed'})
    
    @socketio.on('unsubscribe')
    def handle_unsubscribe(data):
        """Handle unsubscription from ticker updates"""
        try:
            from flask import session
            user_id = session.get('user_id')
            if not user_id:
                return
            
            ticker = data.get('ticker', '').upper()
            if not ticker:
                return
            
            room_name = f"ticker_{ticker}"
            
            # Leave the room
            leave_room(room_name)
            
            # Remove from subscriptions
            if room_name in active_subscriptions and user_id in active_subscriptions[room_name]:
                del active_subscriptions[room_name][user_id]
                if not active_subscriptions[room_name]:
                    del active_subscriptions[room_name]
            
            print(f"User {user_id} unsubscribed from {ticker}")
            emit('unsubscribed', {'ticker': ticker, 'status': 'success'})
            
        except Exception as e:
            print(f"Unsubscribe error: {e}")
    
    return socketio

def start_background_thread():
    """Start the background thread for real-time data updates"""
    global background_thread, background_thread_started
    
    if not background_thread_started:
        background_thread_started = True
        background_thread = threading.Thread(target=background_task)
        background_thread.daemon = True
        background_thread.start()
        print("Background thread started for real-time data updates")

def background_task():
    """Background task to fetch and broadcast real-time data"""
    while True:
        try:
            if not active_subscriptions:
                time.sleep(5)  # No active subscriptions, wait longer
                continue
            
            # Get all unique tickers from active subscriptions
            tickers = set()
            for room_name in active_subscriptions.keys():
                if room_name.startswith('ticker_'):
                    ticker = room_name.replace('ticker_', '')
                    tickers.add(ticker)
            
            # Fetch data for all subscribed tickers
            for ticker in tickers:
                try:
                    room_name = f"ticker_{ticker}"
                    if room_name in active_subscriptions and active_subscriptions[room_name]:
                        # Fetch current data
                        data = fetch_stock_data(ticker)
                        
                        # Broadcast to all subscribers in the room
                        socketio.emit('ticker_update', {
                            'ticker': ticker,
                            'data': data,
                            'timestamp': datetime.now().isoformat()
                        }, room=room_name)
                        
                except Exception as e:
                    print(f"Error updating {ticker}: {e}")
            
            # Wait 30 seconds before next update
            time.sleep(30)
            
        except Exception as e:
            print(f"Background task error: {e}")
            time.sleep(10)

def broadcast_price_alert(user_id, ticker, condition, target_price, current_price):
    """Broadcast price alert to specific user"""
    try:
        if socketio:
            socketio.emit('price_alert', {
                'ticker': ticker,
                'condition': condition,
                'target_price': target_price,
                'current_price': current_price,
                'message': f"{ticker} has {condition} ${target_price}",
                'timestamp': datetime.now().isoformat()
            }, room=f"user_{user_id}")
    except Exception as e:
        print(f"Error broadcasting price alert: {e}")

def broadcast_market_update(data):
    """Broadcast general market updates to all connected users"""
    try:
        if socketio:
            socketio.emit('market_update', {
                'data': data,
                'timestamp': datetime.now().isoformat()
            }, broadcast=True)
    except Exception as e:
        print(f"Error broadcasting market update: {e}")

# Export the init function
__all__ = ['init_socketio', 'broadcast_price_alert', 'broadcast_market_update']

